src/main/java/uk/gov/justice/digital/hmpps/whereabouts/services/vlboptionsfinder/VideoLinkBookingOptionsFinder.kt

```diff
@@ -0,0 +1,160 @@
+import uk.gov.justice.digital.hmpps.whereabouts.dto.prisonapi.ScheduledAppointmentDto
+import uk.gov.justice.digital.hmpps.whereabouts.services.locationfinder.Interval
+import java.time.LocalTime
+import java.util.TreeMap
+
+sealed class Event(val time: LocalTime)
+class StartEvent(time: LocalTime) : Event(time)
+class EndEvent(time: LocalTime) : Event(time)
+
+/**
+ * Takes a list of Events representing the start and end of appointments during a day.
+ * Uses this information to answer the question 'is the Timeline available (free of appointments) during this interval'
+ * (isFreeForInterval(interval: Interval) below.
+ */
+class Timeline(events: List<Event>) {
+
+  /*
+    The periods of time when there are no appointments.
+    The keys are the times at which an empty period started
+    while the values are the corresponding times when the empty period ended.
+   */
+  private val emptyPeriods = TreeMap<LocalTime, LocalTime>()
+
+  init {
+    // The number of bookings in this room.
+    var currentBookings = 0
+
+    // The time of the previous event
+    var previousEventTime: LocalTime = LocalTime.MIN
+
+    // The time when the room last became unoccupied.
+    // This will be null when the room is currently occupied.
+    var freePeriodStarted: LocalTime? = LocalTime.MIN
+
+    fun updateStateForPreviousEventTime() {
+      when (currentBookings) {
+        0 -> {
+          // The room became unoccupied or continued to be unoccupied
+          if (freePeriodStarted == null) {
+            // The room became unoccupied. Remember the time at which this happened
+            freePeriodStarted = previousEventTime
+          }
+        }
+        else -> {
+          // The room became occupied or continued to be occupied at previousEventTime
+          if (freePeriodStarted != null) {
+            // The room became occupied at previousEventTime.  Record the empty period that ended at previousEventTime
+            emptyPeriods[freePeriodStarted!!] = previousEventTime
+            freePeriodStarted = null
+          }
+        }
+      }
+    }
+
+    events
+      .plus(StartEvent(LocalTime.MAX))
+      .sortedBy { it.time }
+      .forEach {
+        if (it.time > previousEventTime) {
+          /**
+           * More than one event can occur at a given time:
+           * One booking may end at the same time that another begins or there may be overlapping bookings.
+           *
+           * If the time of the current event is greater than the time of the previous event then
+           * all events for that earlier time have been seen and we can make decisions about
+           * that previousEventTime (not the time of the current event):
+           */
+          updateStateForPreviousEventTime()
+        }
+        // Now update state from the current event.
+        currentBookings += when (it) {
+          is StartEvent -> 1
+          is EndEvent -> -1
+        }
+        previousEventTime = it.time
```

Comment:
Might find it interesting, I remembered I had a go at [writing an implementation](https://gist.github.com/andrewrlee/681f62f9cb27261e899bc79649d842f9) of this about 6 months ago, when we first started looking at availability which used a similar approach.

I keyed the timeline based on number of minutes of the day each block started and then flattened the appointments 

The general bit for creating a time line and modifying it for each booking: 
```
fun createTimeLine(startOfDay: LocalTime, endOfDay: LocalTime, interval: Long): TimeLine {
    val indices = (startOfDay.minutesOfDay()..endOfDay.minutesOfDay() step interval)
    return indices.map { Pair(it, mutableListOf<Room>()) }.toMap(TreeMap())
}

fun TimeLine.add(booking: RoomBooking) = (booking.start until booking.end).forEach { slot ->
    this.computeIfPresent(slot) { _, list -> list.also { it.add(booking.room) } }
}
```

Then could make a request for a given time range and see the rooms that were booked for a given time:

```
fun TimeLine.getBookedRoomsBetween(timeRange: TimeRange) = this.subMap(timeRange.start, timeRange.end).flatMap { it.value }.toSortedSet()

fun TimeLine.isAvailable(booking: RoomBooking) = booking.room !in this.getBookedRoomsBetween(booking.timeRange)
```

Reply:
Using the map to sort the appointments is interesting. Looks like you create a map entry for each possible start time.  Not too bad when they're at 15 minute intervals. 
I suppose I could just add all the events to a map by event time, then work through the map from earliest to latest to find out how the events at each significant time changed availability.  That would be better I think.  Might do something like that if we need another variation later on.